tasks/test_results_processor.py

```diff
@@ -0,0 +1,201 @@
+import logging
+import random
+import zlib
+from copy import deepcopy
+from datetime import datetime
+from io import BytesIO
+from json import loads
+from typing import List
+
+from redis.exceptions import LockError
+from shared.celery_config import test_results_processor_task_name
+from shared.config import get_config
+from shared.yaml import UserYaml
+from sqlalchemy.orm import Session
+from testing_result_parsers import (
+    Testrun,
+    parse_junit_xml,
+    parse_pytest_reportlog,
+    parse_vitest_json,
+)
+
+from app import celery_app
+from database.models import Repository, Upload
+from database.models.reports import Test, TestInstance
+from services.archive import ArchiveService
+from services.yaml import read_yaml_field
+from tasks.base import BaseCodecovTask
+
+log = logging.getLogger(__name__)
+
+
+class TestResultsProcessorTask(BaseCodecovTask, name=test_results_processor_task_name):
+    __test__ = False
+
+    async def run_async(
+        self,
+        db_session,
+        *,
+        repoid,
+        commitid,
+        commit_yaml,
+        arguments_list,
+        report_code=None,
+        **kwargs,
+    ):
+        commit_yaml = UserYaml(commit_yaml)
+
+        repoid = int(repoid)
+        log.info(
+            "Received upload test result processing task",
+            extra=dict(repoid=repoid, commit=commitid),
+        )
+
+        testrun_list = []
+        upload_list = []
+
+        # process each report session's test information
+        for args in arguments_list:
+            upload_obj: Upload = (
+                db_session.query(Upload).filter_by(id_=args.get("upload_pk")).first()
+            )
+
+            try:
+                parsed_testruns: List[Testrun] = self.process_individual_arg(
+                    upload_obj, upload_obj.report.commit.repository
+                )
+            except Exception:
+                log.warning(f"Error parsing testruns")
+                return {"successful": False}
+
+            # concat existing and new test information
+            testrun_list += parsed_testruns
+
+            upload_list.append(upload_obj)
+
+        # save relevant stuff to database
+        self.save_report(db_session, testrun_list, upload_obj)
+
+        # kick off notification task stuff
+
+        if self.should_delete_archive(commit_yaml):
+            repository = (
+                db_session.query(Repository)
+                .filter(Repository.repoid == int(repoid))
+                .first()
+            )
+            self.delete_archive(
+                commitid, repository, commit_yaml, uploads_to_delete=upload_list
+            )
+
+        return {
+            "successful": True,
+            "testrun_list": [self.testrun_to_dict(t) for t in testrun_list],
+        }
+
+    def save_report(
+        self, db_session: Session, testrun_list: List[Testrun], upload: Upload
+    ):
+        for testrun in testrun_list:
+            test = (
+                db_session.query(Test)
```

Comment:
This is an n+1 query. Query for the list of tests outside the loop. Then check for existence in the loop.